common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ISmartcardSession.java

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.webview.challengehandlers;
+
+import androidx.annotation.NonNull;
+
+import java.security.PrivateKey;
+import java.util.List;
+
+/**
+ * An abstract session that handles pulling information from a smartcard.
+ */
+public interface ISmartcardSession {
+
+    /**
+     * Gets a List of ICertDetails based on certificates stored on the smartcard.
+     * @return a List of ICertDetails. If there are no certificates found, an empty List should be returned.
+     * @throws Exception that could vary upon smartcard type.
+     */
+    @NonNull
+    List<ICertDetails> getCertDetailsList() throws Exception;
+
+    /**
+     * Verifies PIN attempt provided by user.
+     * @param pin Char array with pin.
+     * @return true if PIN is verified; false otherwise.
+     * @throws Exception that could vary upon smartcard type.
+     */
+    boolean verifyPin(char[] pin) throws Exception;
+
+    /**
+     * Gets number of PIN attempts remaining for smartcard.
+     * @return number of PIN attempts remaining.
+     * @throws Exception that could vary upon smartcard type.
+     */
+    int getPinAttemptsRemaining() throws Exception;
+
+    /**
+     * Instantiate a PrivateKey from the chosen certificate that can be passed to a ClientCertRequest proceed method for authentication.
+     * @param certDetails ICertDetails of chosen certificate.
+     * @param pin Char array containing verified PIN.
+     * @return a PrivateKey.
+     * @throws Exception that could vary upon smartcard type.
+     */
+    @NonNull
+    PrivateKey getKeyForAuth(ICertDetails certDetails, char[] pin) throws Exception;
```

Comment:
Does the `PrivateKey` enter the apps memory here? That seems kinda weird.... I would have thought it was not exposed to the app itself, like a TPM behaves

Reply:
From my understanding, the `PrivateKey` being obtained from YubiKit (`PivPrivateKey`, in their case) isn't an actual private key, but there are methods within the object which allow it to perform private key operations without removing the private key from the smartcard. YubiKit includes SPI implementations that call these methods after `proceed` is called on `ClientCertRequest`.

Reply:
But I can see how the name `getKeyForAuth`, which returns a `PrivateKey`, can give off the wrong impression here.... [`proceed`](https://developer.android.com/reference/android/webkit/ClientCertRequest#proceed(java.security.PrivateKey,%20java.security.cert.X509Certificate[])) requires a `PrivateKey` object, but not necessarily a private key. Do you have any renaming suggestions?

Reply:
@iambmelt The PrivateKey here doesn't contain the raw (encoded) key (if you try to use it with other non-YubiKit Crypto implementations (i.e. the BouncyCastle one), it would fail).

(PivPrivateKey actually contains rawSignOrDecrypt() function)

Reply:
This all makes sense... Most TPMs have similar _looking_ interfaces where it appears that the `PrivateKey` enters memory but in reality what you get is an empty proxy object that behaves like a PrivateKey but doesn't introduce any key material into app memory space.

Comments here can help, but LGTM -- it makes sense as is